src/otx/v2/adapters/torch/lightning/modules/models/visual_prompters/segment_anything.py

```diff
@@ -0,0 +1,831 @@
+    #     forward for training/validation/prediction     #
+    ######################################################
+    def forward_train(
+        self,
+        images: Tensor,
+        bboxes: list[Tensor],
+        points: tuple[Tensor, Tensor] | None = None,
+        masks: Tensor | None = None,
+    ) -> tuple[list[Tensor], list[Tensor]]:
+        """Forward method for SAM training/validation/prediction.
+
+        Args:
+            images (Tensor): Images with shape (B, C, H, W).
+            bboxes (List[Tensor]): A Nx4 array given a box prompt to the model, in XYXY format.
+            points (Tuple[Tensor, Tensor], optional): Point coordinates and labels to embed.
+                Point coordinates are BxNx2 arrays of point prompts to the model.
+                Each point is in (X,Y) in pixels. Labels are BxN arrays of labels for the point prompts.
+                1 indicates a foreground point and 0 indicates a background point.
+            masks (Optional[Tensor], optional): A low resolution mask input to the model, typically
+                coming from a previous prediction iteration. Has form Bx1xHxW, where
+                for SAM, H=W=256. Masks returned by a previous iteration of the
+                predict method do not need further transformation.
+
+        Returns:
+            pred_masks (List[Tensor]): List with predicted masks with shape (B, 1, H, W).
+            ious (List[Tensor]): List with IoU predictions with shape (N, 1).
+        """
+        image_embeddings = self.image_encoder(images)
+        pred_masks = []
+        ious = []
+        for embedding, bbox in zip(image_embeddings, bboxes):
+            sparse_embeddings, dense_embeddings = self.prompt_encoder(
+                points=points,
+                boxes=bbox,
+                masks=masks,
+            )
+
+            low_res_masks, iou_predictions = self.mask_decoder(
+                image_embeddings=embedding.unsqueeze(0),
+                image_pe=self.prompt_encoder.get_dense_pe(),
+                sparse_prompt_embeddings=sparse_embeddings,
+                dense_prompt_embeddings=dense_embeddings,
+                multimask_output=False,  # when given multiple prompts. if there is single prompt True would be better.
+            )
+
+            pred_masks.append(low_res_masks)
+            ious.append(iou_predictions)
+
+        return pred_masks, ious
+
+    def training_step(self, batch: dict, batch_idx: int) -> Tensor:
+        """Training step for SAM.
+
+        Args:
+            batch (Dict): Batch data.
+            batch_idx (int): Batch index.
+
+        Returns:
+            loss (Tensor): Loss tensor.
+        """
+        _ = batch_idx
+        images = batch["images"]
+        bboxes = batch["bboxes"]
+        points = batch["points"]
+        gt_masks = batch["gt_masks"]
+
+        pred_masks, iou_predictions = self.forward_train(images, bboxes, points)
+
+        loss_dice = 0.0
```

Comment:
As I mentioned above, this loss computation here does not seem to be Lightning friendly. It could potentially be a lot simpler if we have a loss component, which Lightning can automatically update.

Reply:
For example, refer to this loss example 
https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/models/dsr/loss.py